Azure/sonic-buildimage
# [build] Add option to avoid Docker base image :latest tag

* Define slave_base_tag_ref variable in Makefile.work containing
  specific base image tag to use, rather than always defaulting
  to :latest.

* Add an ARG command before FROM statement in Dockerfile.user
  for sonic-slave and sonic-slave-stretch. ARG variable defaults
  to latest if slave_base_tag_ref not specified in Makefile.work.

The presumption to always refer to the :latest tagged Docker base
image when creating the user image causes problems in a shared
build server environment, where the most recently created base
image (i.e. the current :latest tag) may not be compatible with
the current build. For example, different users working in
different branches may all be sharing the same build server.

Signed-off-by: Greg Paussa <greg.paussa@broadcom.com>



**- What I did**
Added a DOCKER_AVOID_BASE_TAG_LATEST build option to rules/config that forces the Docker user image creation to refer to its base image by a specific tag rather than rely on the :latest tag. This is needed in a shared build server environment where builds from different developers and/or different SONiC branches all converge on the same Docker daemon instance running on the build server. The :latest tag is always assigned to the most recent base image built, which might not correspond to the base image needed for a particular build, thus causing various build errors that mostly manifest as missing Debian packages or package version mismatches.

**NOTE TO REVIEWERS:** This PR relies on Docker support of "ARG before FROM," which was first introduced in Docker version [17.05.1-ce](https://docs.docker.com/engine/release-notes/#17050-ce). Although there is no mention of a minimum required Docker version for the build server in the SONiC Building Guide pages, please consider whether it is reasonable to assume that Docker 17.05.1-ce or later must be used for SONiC build hosts before approving this PR.

**- How I did it**
Added an ARG before the FROM statement at the top of the sonic-slave/Dockerfile.user and sonic-slave-stretch/Dockerfile.user files. The ARG variable defaults to latest, but can be overridden in Makefile.work to reference the SLAVE_BASE_TAG so that it refers to the specific, matching base image for the build. This override is activated by un-commenting the DOCKER_AVOID_BASE_TAG_LATEST = y line in rules/config.

**- How to verify it**
A mismatch in the base image tag causes an error during the user image creation, which would be obvious to the developer. To verify a successful build of the user image, check its Docker history to see that the right base image ID is referenced, as illustrated by the following example:

<pre>
Hint: Focus on <b>bold</b> values in this output.

# make showtag
+++ Making showtag +++
BLDENV=stretch make -f Makefile.work showtag
make[1]: Entering directory '/builds/sonic/someuser/repos/sonic-buildimage'
sonic-slave-stretch-someuser:ed01ec3b145
sonic-slave-stretch-base:<b>93e3a1c4b07</b>
make[1]: Leaving directory '/builds/sonic/someuser/repos/sonic-buildimage'

# docker images | egrep "sonic-slave-stretch-someuser|REPOSITORY"
REPOSITORY                                     TAG                 IMAGE ID            CREATED             SIZE
sonic-slave-stretch-someuser                   ed01ec3b145         692723e02f00        About an hour ago   4.16GB

# docker images | egrep "sonic-slave-stretch-base|REPOSITORY"
REPOSITORY                                     TAG                 IMAGE ID            CREATED             SIZE
sonic-slave-stretch-base                       b501a9391d2         384ac9293e94        7 days ago          4.15GB
sonic-slave-stretch-base                       cd636ae59de         0433427b4850        8 days ago          4.3GB
sonic-slave-stretch-base                       3e6941322e5         e06c913f4263        11 days ago         4.17GB
sonic-slave-stretch-base                       <b>93e3a1c4b07</b>         <b>19a7e622da3d</b>        11 days ago         4.16GB
sonic-slave-stretch-base                       9324698b31e         1faf09d99eee        12 days ago         4.16GB
sonic-slave-stretch-base                       b12f3e165ce         1d5f3588a83e        12 days ago         4.3GB
sonic-slave-stretch-base                       latest              1d5f3588a83e        12 days ago         4.3GB
sonic-slave-stretch-base                       8f70fa10dae         7d220a420831        2 weeks ago         4.16GB
sonic-slave-stretch-base                       efee9dd0e26         00663d432f28        2 weeks ago         4.15GB

# docker history 692723e02f00
IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
692723e02f00        About an hour ago   /bin/sh -c #(nop)  USER someuser                0B                         
1711748a2f3d        About an hour ago   |4 guid=841 hostname=sonic-bldsrvr-2 uid=7â¦     701B                      
d87d8799c2b3        About an hour ago   |4 guid=841 hostname=sonic-bldsrvr-2 uid=7â¦     398B                      
d0ade02c8a82        About an hour ago   |4 guid=841 hostname=sonic-bldsrvr-2 uid=7â¦     398B                      
57c29a0d6bd7        About an hour ago   /bin/sh -c #(nop) COPY file:551f41e8b32096c1â¦   398B                      
c8b473835fe8        About an hour ago   |4 guid=841 hostname=sonic-bldsrvr-2 uid=7â¦     54B                       
186483e78c95        About an hour ago   |4 guid=841 hostname=sonic-bldsrvr-2 uid=7â¦     24B                       
cf35220b7742        About an hour ago   |4 guid=841 hostname=sonic-bldsrvr-2 uid=7â¦     2.52kB                    
7cd978c6cc08        About an hour ago   |4 guid=841 hostname=sonic-bldsrvr-2 uid=7â¦     10kB                      
3a2584a6f546        About an hour ago   |4 guid=841 hostname=sonic-bldsrvr-2 uid=7â¦     2.47kB                    
c4ea8ffe9efe        About an hour ago   /bin/sh -c #(nop)  ENV USER=someuser            0B                         
829ee3b3af2c        About an hour ago   /bin/sh -c #(nop)  ENV BUILD_HOSTNAME=sonic-â¦   0B                        
a191fb031dda        About an hour ago   /bin/sh -c #(nop)  ARG hostname                 0B                         
b1949b721d79        About an hour ago   /bin/sh -c #(nop)  ARG guid                     0B                         
e4ec477ad1b7        About an hour ago   /bin/sh -c #(nop)  ARG uid                      0B                         
ac52e018e0e2        About an hour ago   /bin/sh -c #(nop)  ARG user                     0B                         
<b>19a7e622da3d</b>        11 days ago         |2 debian_archive_mirror= debian_org_mirror=â¦   704B                      
<missing>           11 days ago         |2 debian_archive_mirror= debian_org_mirror=â¦   117MB                     
<missing>           11 days ago         |2 debian_archive_mirror= debian_org_mirror=â¦   670kB                     
---snip---
</pre>

**- Description for the changelog**

Add optional build configuration variable to avoid relying on :latest tag when identifying Docker base build image.

**- A picture of a cute animal (not mandatory but encouraged)**
